server/src/handlers.py

```
@@ -1,39 +1,104 @@
-    limit = int(request.query.get('limit', '10'))
-    rooms, total_count = await db.get_rooms(page, limit)
-    return web.json_response({
-        "rooms": rooms,
-        "paginator": {
-            "page": page,
-            "limit": limit,
-            "total_pages": math.ceil(total_count / limit),
-        }
-    }, status=200, dumps=parse_json)
+    except UniqueViolationError:
+        return web.json_response({'message': 'user with that name already exists'}, status=409)
+    return web.json_response({'message': 'OK'}, status=201)
+
+
+async def login(request: web.Request) -> web.Response:
+    data = await request.json()
+    remember = data.pop('remember', False)
+    user = User(data)
+    try:
+        user.validate()
+    except DataError as e:
+        return web.json_response(e.to_primitive(), status=400)
+    stored_user = await db.get_user(request.app['pool'], user)
+    if stored_user is None:
+        return web.json_response({'message': 'user don\'t exists'}, status=404)
+    if not is_compare_messages(stored_user.password, user.password):
+        return web.json_response({'message': 'wrong password'}, status=400)
+    response = web.json_response({'message': 'OK'}, status=200)
+    response.set_cookie(
+        name='token',
+        value=encrypt_jwt(id=stored_user.id, username=stored_user.username),
+        httponly=True,
+        max_age=None if remember else 3600 * 24 * 7
+    )
+    return response
+
+
+async def logout(request: web.Request) -> web.Response:
+    response = web.json_response({'message': 'OK'})
+    response.del_cookie(name='token')
+    return response
+
+
+@auth_required
+async def create_game(request: web.Request) -> web.Response:
+    game = await db.create_game(request.app['pool'], request.user)
+    return web.json_response(asdict(game), status=201)
+
+
+@auth_required
+async def login_game(request: web.Request) -> web.Response:
+    data = await request.json()
+    if not is_numeric(request.match_info['gID']):
```

1) How do you know that game with that id is not exists?
2) Check state of game that you take by id, request body will be empty

---

1 if id is not a number, then you need to return an error, since we have only an integer id and why check further if the id is already wrong?

---

Then add regular expression to the path, like u did in the `get_game` handler